### ISSUE TITLE
How to interpret fact_check and comparaison stats ? 

### ISSUE THREAD

**Issue by carlosandrea** [llmware-ai/llmware#114]
Hello, 

I just ran dragon_rag_benchmark_tests_llmware.py, on llmware/dragon-red-pajama-7b-v0 
I'm not sure how to interpret Fact Check stats and comparison stats.
->For some cases, Gold and LLM answers are equivalent but stats are low :

"12. llm_response -  Delta Inc. and ABC Company
12. gold_answer - ABC Company and Delta Inc.
12. time_taken - 0.16
12. comparison_stats - {'percent_display': '66.7%', 'confirmed_words': ['delta', 'company'], 'unconfirmed_words': ['abc'], 'verified_token_match_ratio': 0.6666666666666666, 'key_point_list': [{'key_point': ' Delta Inc. and ABC Company', 'entry': 0, 'verified_match': 0.6666666666666666}]}
12. source - 0 {'text': 'Company 8675 River Run Road Marlington CT 09392 INVOICE TO Delta Inc 8723 Wilton Boulevard Los Angeles CA HEBE HERACLES\tLASER ', 'match_score': 0.75, 'source': 'NA', 'page_num': 'NA'}"

->For some cases Gold and LLM answers are not equivalent but stats are very high :

"54. llm_response -  Section 18(a), which is governed by the laws of the State of Delaware.
54. gold_answer - Section 19
54. time_taken - 0.42
54. comparison_stats - {'percent_display': '100.0%', 'confirmed_words': ['section', '18a', 'governed', 'laws', 'state', 'delaware'], 'unconfirmed_words': [], 'verified_token_match_ratio': 1.0, 'key_point_list': [{'key_point': ' Section 18(a), which is governed by the laws of the State of Delaware.', 'entry': 0, 'verified_match': 1.0}]}
54. source - 0 {'text': 'and interpreted under the laws of the State of Massachusetts except with respect to Section 18(a) of this Agreement which ', 'match_score': 1.0, 'source': 'NA', 'page_num': 'NA'}"

"107. llm_response -  Q2 of calendar 2024.
107. gold_answer - Not Found.
107. time_taken - 0.2
107. fact_check - 0 {'fact': '2024', 'status': 'Confirmed', 'text': ' ... Outlook NVIDIA??s outlook for the third quarter of fiscal 2024 is as follows:Revenue is expected to be $16.00 billion,  ... ', 'page_num': 'NA', 'source': 'NA'}
107. comparison_stats - {'percent_display': '100.0%', 'confirmed_words': ['q2', 'calendar', '2024'], 'unconfirmed_words': [], 'verified_token_match_ratio': 1.0, 'key_point_list': [{'key_point': ' Q2 of calendar 2024.', 'entry': 0, 'verified_match': 1.0}]}
107. source - 0 {'text': 'a secondgeneration version with HBM3e memory expected to ship in Q2 of calendar 2024 ', 'match_score': 1.0, 'source': 'NA', 'page_num': 'NA'}

"
Also, i'm not sure to understand fact_check, In answer 107..
Maybe I'm missing something..

**Comment by doberst** [comment]
@carlosandrea:   thanks for your question.   Here is a quick explanation of each of the fact checks that are being run and what they are testing:

1.  Numbers "Fact Check" - this will review the LLM response output, and look for tokens that appear to be a number value, and then compare with the evidence provided, and look for confirmation that the numerical value can be found in the source materials.   If found, then status is "Confirmed" and the corresponding text snippet is provided.   The function attempts to match value, even if formulated slightly differently, e.g., "$100.00" should match "100" or "100.00", etc.

2.  Comparison Stats - this function tokenizes the LLM response output, and looks for each of those tokens in the corresponding context evidence.   "Confirmed words" are words in the LLM response that are confirmed to be found in the context evidence.  "Unconfirmed words" are words in the LLM response that are not matched by the context passage. The percent display and verified match are based on the words in the LLM response, so if the LLM response has three words, and all three words are found in the evidence, then it will be 1.0 or 100%.

3.  Source Review - using a token comparison between the LLM response output and the context evidence, the function attempts to extract statistically high-matching text snippets that were the basis of the LLM's response.

Hope this clarifies how the functions are intended to work.   

Appreciate you sharing the individual screen shot responses -> we will review each of them in detail and revert back...

**Comment by doberst** [comment]
@carlosandrea - yes, to answer (belatedly), we did a manual review to prepare the test scores for each model.  The tools are useful to help accelerate, but in the case of preparing the scores for the Dragon and BLING models, we supplemented with a manual review.   If you look in the files for each model (in the corresponding HuggingFace repository), you will see the actual scoring sheet and results for each model.
